rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,126 @@
+- Kahan summation algorithm will be used
+
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero, 
+old formula will be used.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+- If needed, may be kept confidential and not disclosed to external customes 
+until tested internally
+
+Cons:
+- No fine-tuning per instance of bnorm primitive
+- Harder to reproduce bugs related to bnorm: value of this flag will need to be
+ reported in bug reports to allow reproduction.
+ - Inconvenient long-term. State of oneDNN will have a dependency which is not
+ obvious to users.
+
+### Enable through additional flag in bnorm primitive descriptor
+Bnorm descriptor has flags field `dnnl_normalization_flags_t`. Add new flag 
+`dnnl_one_pass = 0x20u`. When set to non-zero, new formula will be used.
+
+Pros:
+- Fine-tuning per instance of bnorm primitive
+- Suitable to become permanent solution, should both formulas be kept in oneDNN
+
+Cons:
+- API change needed.
+- In case a customer was passing garbage into bit5 of flags, the default
+state is undefined and behavior may unexpectedly change.
+- If we decide to remove it later and keep only one formula, it will require
+another API change; it will leave behind a deprecated flag.
+- Inconsistency: not all implementations of bnorm primitive support that flag.
+Specifically, CPU implementation doesn't have it.
+- Inconvenient for validation. In order to test this feature with a topology,
+code change in frameworks in needed as frameworks will need to set this flag 
+for each bnorm instance.
+
+### Enable through new API call
+Introduce new API function `void set_bnorm_one_pass_mode(bool onepass)`. This
+function will set global variable which will subsequently control behavior of
+all instances of bnorm primitive.
+
+Pros:
+- API stays backwards-compatible
+- Clearly defined default state
+
+Cons:
+- API change needed.
+- No fine-tuning per instance of bnorm primitive
+- If we decide to remove it later and keep only  one formula, it will require
+ another API change; it will leave behind a deprecated function.
+
+## Validation
+New formula was tested with RN50 and it trained to SotA. Further tests with
+broader set of topologies are needed. Until such tests are performed, new
+formula should be optionally available but disabled by default. The point
+of this request is to make new formula available for further validation, while
+external customers keep using the old and proven formula.
```

Comment:
All in all I would vote for the option with an additional flag, because this way we define some API, which can be used by our tests and user. On the other side since we are not sure it will be widely used we minimize amount of changes on our side to just a single additional flag.